File: Data Structures & Algorithms/is-palindrome/submission-14.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        l = 0
        r = len(s) - 1

        while l <= r:
            left = s[l].lower()
            right = s[r].lower()

            if not self.isCharValid(left):
                l += 1
                continue

            if not self.isCharValid(right):
                r -= 1
                continue

            if left != right:
                return False

            r -= 1
            l += 1

        return True

    def isCharValid(self, c):
        return 'a' <= c <= 'z' or '0' <= c <= '9'

File: Data Structures & Algorithms/binary-tree-diameter/submission-14.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root: TreeNode | null): number {
        let maxDiam = 0

        function maxDepth(node: TreeNode | null) {
            if (!node) return 0

            const left = maxDepth(node.left)
            const right = maxDepth(node.right)

            maxDiam = Math.max(maxDiam, left + right)

            return 1 + Math.max(left, right)
        }

        maxDepth(root)
        return maxDiam
    }
}
